File: zero_war.c
/*
 * ╔══════════════════════════════════════════════════════════════╗
 * ║                    ⚔  ZERO WAR  ⚔                           ║
 * ║              Your Zero — C Edition                          ║
 * ╚══════════════════════════════════════════════════════════════╝
 *
 * SETUP (Linux / WSL / Raspberry Pi):
 *   sudo apt install libmosquitto-dev
 *   gcc zero_war.c -o zero_war -lmosquitto && ./zero_war
 *
 * SETUP (macOS):
 *   brew install mosquitto
 *   gcc zero_war.c -o zero_war -lmosquitto && ./zero_war
 *
 * WHAT THIS SCRIPT DOES:
 *   1. Connects to the class MQTT broker over TLS
 *   2. Publishes your zero's status every 10 seconds
 *   3. Listens for incoming attacks and prints them
 *   4. Lets you attack another zero by typing their ID
 *
 * YOUR MISSION:
 *   Get it compiling and running, then improve it:
 *   - [ ] Track HP properly and update it when attacked
 *   - [ ] Print a bar like [########--] for HP
 *   - [ ] Add a cooldown between attacks (use time())
 *   - [ ] Print "GAME OVER" and exit when HP hits 0
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <mosquitto.h>

/* ── CONFIGURATION ─────────────────────────────────────────────────────────
   Fill in your details before compiling! */

#define BROKER    "3db5059fc43148f4b17376a4bd9f28f2.s1.eu.hivemq.cloud"
#define PORT      8883
#define USERNAME  "YOUR_USERNAME"
#define PASSWORD  "YOUR_PASSWORD"

#define MY_ID     "c01"        /* Pick a unique short ID */
#define MY_NAME   "YourName"   /* Your display name      */
#define MY_TEAM   "A"          /* "A" or "B"             */

/* ── GAME STATE ────────────────────────────────────────────────────────────*/
int hp          = 100;
int game_active = 0;

/* We need a pointer to the mosquitto client in callbacks,
   so we store it globally here */
struct mosquitto *mosq = NULL;


/* ── HELPERS ───────────────────────────────────────────────────────────────*/

/* Build a simple JSON string without an external library.
   In real projects you'd use cJSON or json-c.
   buf must be large enough to hold the result. */
void build_status_json(char *buf, int bufsize) {
    snprintf(buf, bufsize,
        "{\"id\":\"%s\",\"name\":\"%s\",\"team\":\"%s\","
        "\"hp\":%d,\"alive\":%s}",
        MY_ID, MY_NAME, MY_TEAM,
        hp, (hp > 0) ? "true" : "false"
    );
}

/* Parse a single integer field from a flat JSON string.
   e.g. get_json_int("{\"damage\":15}", "damage")  →  15
   This is a simple approach — for complex JSON use cJSON. */
int get_json_int(const char *json, const char *key) {
    /* Look for "key": in the string */
    char search[64];
    snprintf(search, sizeof(search), "\"%s\":", key);

    const char *pos = strstr(json, search);
    if (!pos) return -1;

    /* Move past the key and colon to the value */
    pos += strlen(search);
    return atoi(pos);
}

/* Same idea, but extracts a string field into out_buf */
void get_json_str(const char *json, const char *key, char *out_buf, int out_size) {
    char search[64];
    snprintf(search, sizeof(search), "\"%s\":\"", key);

    const char *pos = strstr(json, search);
    if (!pos) { out_buf[0] = '\0'; return; }

    pos += strlen(search);
    int i = 0;
    while (*pos && *pos != '"' && i < out_size - 1)
        out_buf[i++] = *pos++;
    out_buf[i] = '\0';
}

/* Publish our current HP and info to the scoreboard topic */
void publish_status(void) {
    char topic[64];
    char payload[256];

    snprintf(topic, sizeof(topic), "zerowars/status/%s", MY_ID);
    build_status_json(payload, sizeof(payload));

    /* retain=true means the broker keeps this message and sends it
       to any new subscriber immediately — great for status */
    mosquitto_publish(mosq, NULL, topic, strlen(payload), payload, 0, true);
    printf("📤 Status published: HP=%d\n", hp);
}

/* Attack a target zero by their ID */
void send_attack(const char *target_id) {
    char topic[64];
    char payload[128];

    /* Random damage 10–25 */
    int damage = 10 + (rand() % 16);

    snprintf(topic,   sizeof(topic),   "zerowars/attack/%s", target_id);
    snprintf(payload, sizeof(payload), "{\"from\":\"%s\",\"damage\":%d}", MY_ID, damage);

    mosquitto_publish(mosq, NULL, topic, strlen(payload), payload, 0, false);
    printf("⚡ Attacked %s for %d damage!\n", target_id, damage);
}


/* ── CALLBACKS ─────────────────────────────────────────────────────────────
   These are called automatically by the mosquitto library.
   You register them with mosquitto_*_callback_set() in main(). */

/* Called when connection to broker is established (or fails) */
void on_connect(struct mosquitto *m, void *userdata, int rc) {
    if (rc == 0) {
        printf("✅ Connected to HiveMQ as '%s' [%s]\n", MY_NAME, MY_ID);

        /* Subscribe to our attack topic and the teacher broadcast */
        char attack_topic[64];
        snprintf(attack_topic, sizeof(attack_topic), "zerowars/attack/%s", MY_ID);

        mosquitto_subscribe(m, NULL, attack_topic,      0);
        mosquitto_subscribe(m, NULL, "zerowars/broadcast", 0);

        printf("👂 Listening on: %s\n", attack_topic);

        /* Announce ourselves to the dashboard */
        publish_status();
    } else {
        /* rc values: 1=wrong protocol, 2=bad client ID, 3=broker unavailable,
                      4=bad credentials, 5=not authorised */
        fprintf(stderr, "❌ Connection failed. Return code: %d\n", rc);
        fprintf(stderr, "   Check USERNAME, PASSWORD, and BROKER.\n");
    }
}

/* Called every time a message arrives on a subscribed topic */
void on_message(struct mosquitto *m, void *userdata,
                const struct mosquitto_message *msg)
{
    /* msg->payload is NOT null-terminated, so we copy it into a proper string */
    char payload[512];
    int  len = msg->payloadlen < (int)sizeof(payload) - 1
                   ? msg->payloadlen
                   : (int)sizeof(payload) - 1;
    memcpy(payload, msg->payload, len);
    payload[len] = '\0';

    char my_attack_topic[64];
    snprintf(my_attack_topic, sizeof(my_attack_topic), "zerowars/attack/%s", MY_ID);

    /* ── Incoming attack ───────────────────────────────────── */
    if (strcmp(msg->topic, my_attack_topic) == 0) {
        int  damage   = get_json_int(payload, "damage");
        char attacker[32];
        get_json_str(payload, "from", attacker, sizeof(attacker));

        if (damage < 0) damage = 10;   /* default if parsing fails */

        hp -= damage;
        if (hp < 0) hp = 0;

        printf("\n💥 HIT by %s! -%d HP  |  HP remaining: %d\n",
               attacker, damage, hp);

        if (hp == 0) {
            printf("💀 YOU HAVE BEEN ELIMINATED. Better luck next round!\n");
        }

        /* TODO (your task): publish updated status after being hit */
        /* publish_status(); */
    }

    /* ── Teacher broadcast ─────────────────────────────────── */
    else if (strcmp(msg->topic, "zerowars/broadcast") == 0) {
        char cmd[32];
        get_json_str(payload, "cmd", cmd, sizeof(cmd));

        if (strcmp(cmd, "START") == 0) {
            hp          = 100;
            game_active = 1;
            printf("\n🚀 === GAME STARTED! === Fight!\n");
            publish_status();

        } else if (strcmp(cmd, "STOP") == 0) {
            game_active = 0;
            printf("\n🛑 === GAME OVER ===\n");

        } else if (strcmp(cmd, "PING") == 0) {
            printf("📡 PING received — responding...\n");
            publish_status();
        }
    }
}

/* Called when disconnected from broker */
void on_disconnect(struct mosquitto *m, void *userdata, int rc) {
    printf("\n⚠️  Disconnected (rc=%d). Reconnecting...\n", rc);
}


/* ── MAIN ──────────────────────────────────────────────────────────────────*/
int main(void) {
    srand((unsigned)time(NULL));

    /* mosquitto_lib_init() must be called once before anything else */
    mosquitto_lib_init();

    /* Create a new mosquitto client instance.
       Arguments: client ID string, clean_session=true, userdata pointer */
    char client_id[32];
    snprintf(client_id, sizeof(client_id), "zero-%s", MY_ID);
    mosq = mosquitto_new(client_id, true, NULL);

    if (!mosq) {
        fprintf(stderr, "❌ Failed to create mosquitto client\n");
        return 1;
    }

    /* Register callbacks */
    mosquitto_connect_callback_set(mosq,    on_connect);
    mosquitto_message_callback_set(mosq,    on_message);
    mosquitto_disconnect_callback_set(mosq, on_disconnect);

    /* Set credentials */
    mosquitto_username_pw_set(mosq, USERNAME, PASSWORD);

    /* Enable TLS — NULL arguments use system CA certificates */
    int rc = mosquitto_tls_set(mosq, NULL, NULL, NULL, NULL, NULL);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "❌ TLS setup failed: %s\n", mosquitto_strerror(rc));
        return 1;
    }

    /* Connect to the broker.
       keepalive=60 means send a PING every 60 seconds to stay connected */
    printf("🔌 Connecting to %s:%d ...\n", BROKER, PORT);
    rc = mosquitto_connect(mosq, BROKER, PORT, 60);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "❌ Connect error: %s\n", mosquitto_strerror(rc));
        return 1;
    }

    /* Start the network loop in a background thread.
       This handles all send/receive operations automatically. */
    mosquitto_loop_start(mosq);

    /* Give it a moment to connect and trigger on_connect */
    sleep(2);

    /* ── Input loop ─────────────────────────────────────────── */
    printf("\nCommands:\n");
    printf("  attack <target_id>   attack another zero\n");
    printf("  status               print your current HP\n");
    printf("  quit                 exit\n\n");

    char line[128];
    while (1) {
        printf("> ");
        fflush(stdout);

        if (!fgets(line, sizeof(line), stdin)) break;

        /* Strip newline */
        line[strcspn(line, "\n")] = '\0';

        if (strncmp(line, "attack ", 7) == 0) {
            const char *target = line + 7;   /* pointer arithmetic: skip "attack " */
            send_attack(target);

        } else if (strcmp(line, "status") == 0) {
            printf("HP: %d/100  |  Team: %s  |  ID: %s\n", hp, MY_TEAM, MY_ID);

        } else if (strcmp(line, "quit") == 0 || strcmp(line, "q") == 0) {
            printf("👋 Disconnecting...\n");
            break;

        } else if (line[0] != '\0') {
            printf("Unknown command. Try: attack <id> | status | quit\n");
        }
    }

    /* Clean shutdown */
    mosquitto_loop_stop(mosq, true);
    mosquitto_disconnect(mosq);
    mosquitto_destroy(mosq);
    mosquitto_lib_cleanup();

    return 0;
}

File: zero_war.py
"""
╔══════════════════════════════════════════════════════════════╗
║                    ⚔  ZERO WAR  ⚔                           ║
║              Your Zero — Python Edition                      ║
╚══════════════════════════════════════════════════════════════╝

SETUP:
    pip install paho-mqtt

WHAT THIS SCRIPT DOES:
    1. Connects to the class MQTT broker (HiveMQ)
    2. Publishes your zero's status every 10 seconds
    3. Listens for incoming attacks and prints them
    4. Lets you attack another zero by typing their ID

MQTT TOPICS USED:
    zerowars/status/<your_id>        ← you publish here
    zerowars/attack/<your_id>        ← enemies publish here (you listen)
    zerowars/attack/<target_id>      ← you publish here to attack
    zerowars/broadcast               ← teacher sends game commands

YOUR MISSION:
    Get the script running, then improve it:
    - [ ] Track your own HP and print it nicely
    - [ ] Add a cooldown so you can't spam attacks
    - [ ] Make it print a different message depending on how much damage you take
    - [ ] Print a "GAME OVER" message when your HP hits 0
"""

import paho.mqtt.client as mqtt
import json
import time
import threading
import random

# ── CONFIGURATION ─────────────────────────────────────────────────────────────
# Fill in YOUR details here before running!

BROKER   = "3db5059fc43148f4b17376a4bd9f28f2.s1.eu.hivemq.cloud"
PORT     = 8883          # TLS port
USERNAME = "YOUR_USERNAME"
PASSWORD = "YOUR_PASSWORD"

MY_ID    = "p01"         # Pick a unique short ID, e.g. your initials + number
MY_NAME  = "YourName"    # Your display name on the scoreboard
MY_TEAM  = "A"           # "A" or "B" — your teacher will tell you which

# ── GAME STATE ────────────────────────────────────────────────────────────────
hp = 100
game_active = False


# ── CALLBACKS ─────────────────────────────────────────────────────────────────
# These functions are called automatically when MQTT events happen.
# You don't call them yourself — the library calls them for you.

def on_connect(client, userdata, flags, rc, properties=None):
    """Called when the connection to the broker is established."""
    if rc == 0:
        print(f"✅ Connected to HiveMQ as '{MY_NAME}' [{MY_ID}]")

        # Subscribe to topics we want to RECEIVE
        client.subscribe(f"zerowars/attack/{MY_ID}")   # incoming attacks
        client.subscribe("zerowars/broadcast")          # teacher commands
        print(f"👂 Listening on: zerowars/attack/{MY_ID}")

        # Announce ourselves to the game dashboard
        publish_status(client)

    else:
        # rc = return code. Anything other than 0 means something went wrong.
        print(f"❌ Connection failed. Return code: {rc}")
        print("   Check your username, password, and broker address.")


def on_message(client, userdata, msg):
    """Called every time a message arrives on a subscribed topic."""
    global hp, game_active

    topic   = msg.topic
    payload = msg.payload.decode("utf-8")

    # Try to parse the message as JSON
    try:
        data = json.loads(payload)
    except json.JSONDecodeError:
        print(f"[?] Got non-JSON message on {topic}: {payload}")
        return

    # ── Handle incoming attack ────────────────────────────────
    if topic == f"zerowars/attack/{MY_ID}":
        damage    = data.get("damage", 10)
        attacker  = data.get("from", "unknown")

        hp -= damage
        hp  = max(0, hp)   # don't go below 0

        print(f"\n💥 HIT by {attacker}! -{damage} HP  |  HP remaining: {hp}")

        if hp == 0:
            print("💀 YOU HAVE BEEN ELIMINATED. Better luck next round!")

        # TODO (your task): publish your updated status after being hit
        # publish_status(client)

    # ── Handle teacher broadcast ──────────────────────────────
    elif topic == "zerowars/broadcast":
        command = data.get("cmd", "")

        if command == "START":
            hp          = 100
            game_active = True
            print("\n🚀 === GAME STARTED! === Fight!")
            publish_status(client)

        elif command == "STOP":
            game_active = False
            print("\n🛑 === GAME OVER ===")

        elif command == "PING":
            # Teacher is checking who's online
            print("📡 PING received — responding...")
            publish_status(client)


def on_disconnect(client, userdata, rc, properties=None):
    """Called when the connection is lost."""
    print(f"\n⚠️  Disconnected from broker (rc={rc}). Reconnecting...")


# ── PUBLISH HELPERS ───────────────────────────────────────────────────────────

def publish_status(client):
    """Send our current HP and info to the scoreboard."""
    message = {
        "id":     MY_ID,
        "name":   MY_NAME,
        "team":   MY_TEAM,
        "hp":     hp,
        "alive":  hp > 0,
    }
    # json.dumps() converts a Python dict → a JSON string
    payload = json.dumps(message)
    client.publish(f"zerowars/status/{MY_ID}", payload, retain=True)
    print(f"📤 Status published: HP={hp}")


def send_attack(client, target_id):
    """Attack another zero by their ID."""
    damage = random.randint(10, 25)    # random damage between 10 and 25

    message = {
        "from":   MY_ID,
        "damage": damage,
    }
    payload = json.dumps(message)
    client.publish(f"zerowars/attack/{target_id}", payload)
    print(f"⚡ Attacked {target_id} for {damage} damage!")


# ── HEARTBEAT THREAD ──────────────────────────────────────────────────────────
# This runs in the background and sends our status every 10 seconds.
# Threading lets two things happen at the same time: the heartbeat
# AND the input loop below.

def heartbeat_loop(client):
    """Publish status every 10 seconds so the dashboard knows we're alive."""
    while True:
        time.sleep(10)
        publish_status(client)


# ── MAIN ──────────────────────────────────────────────────────────────────────

def main():
    # Create an MQTT client object
    # CallbackAPIVersion.VERSION2 is required for paho-mqtt >= 2.0
    client = mqtt.Client(
        mqtt.CallbackAPIVersion.VERSION2,
        client_id=f"zero-{MY_ID}"
    )

    # Attach our callback functions
    client.on_connect    = on_connect
    client.on_message    = on_message
    client.on_disconnect = on_disconnect

    # Set credentials and enable TLS encryption
    client.username_pw_set(USERNAME, PASSWORD)
    client.tls_set()   # uses system's trusted CA certificates

    # Connect (this does not block — it just starts the process)
    print(f"🔌 Connecting to {BROKER}:{PORT} ...")
    client.connect(BROKER, PORT, keepalive=60)

    # Start the MQTT network loop in a background thread
    # This handles sending/receiving in the background automatically
    client.loop_start()

    # Give it a moment to connect before we start
    time.sleep(2)

    # Start heartbeat in a separate thread
    t = threading.Thread(target=heartbeat_loop, args=(client,), daemon=True)
    t.start()

    # ── Input loop — lets you type commands ───────────────────
    print("\nCommands:")
    print("  attack <target_id>   — attack another zero")
    print("  status               — print your current HP")
    print("  quit                 — exit\n")

    while True:
        try:
            command = input("> ").strip().lower()

            if command.startswith("attack "):
                target = command.split(" ", 1)[1]
                send_attack(client, target)

            elif command == "status":
                print(f"HP: {hp}/100  |  Team: {MY_TEAM}  |  ID: {MY_ID}")

            elif command in ("quit", "exit", "q"):
                print("👋 Disconnecting...")
                break

            else:
                print("Unknown command. Try: attack <id> | status | quit")

        except KeyboardInterrupt:
            break

    client.loop_stop()
    client.disconnect()


if __name__ == "__main__":
    main()
